src/SignalCancellation.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace Amp;
+
+use Revolt\EventLoop;
+
+/**
+ * A SignalCancellation automatically requests cancellation when a given signal is received.
+ */
+final class SignalCancellation implements Cancellation
+{
+    use ForbidCloning;
+    use ForbidSerialization;
+
+    /** @var list<string> */
+    private readonly array $watchers;
+
+    private readonly Cancellation $cancellation;
+
+    /**
+     * @param int|int[] $signals Signal number or array of signal numbers.
+     * @param string $message Message for SignalException. Default is "Operation cancelled by signal".
+     */
+    public function __construct(int|array $signals, string $message = "Operation cancelled by signal")
+    {
+        if (\is_int($signals)) {
+            $signals = [$signals];
+        }
+
+        $this->cancellation = $source = new Internal\Cancellable;
+
+        $trace = null; // Defined in case assertions are disabled.
+        \assert((bool) ($trace = \debug_backtrace(0)));
+
+        $callback = static function () use ($source, $message, $trace): void {
+            if ($trace) {
+                $message .= \sprintf("\r\n%s was created here: %s", self::class, Internal\formatStacktrace($trace));
+            } else {
+                $message .= \sprintf(" (Enable assertions for a backtrace of the %s creation)", self::class);
+            }
+
+            $source->cancel(new SignalException($message));
```

Comment:
What should happen if multiple signals are received?

Reply:
The cancellation is triggered only on the first signal - subsequent signals are ignored. Updated to explicitly cancel all the watchers on the first signal.